rfcs/2020-04-06-1858-automatically-adjust-limits.md

```diff
@@ -0,0 +1,164 @@
+can require considerable trial and error experimentation before a
+satisfactory set of parameters can be achieved. More importantly,
+changes in the service parameters, whether in terms of available
+processing power, bandwidth, or changes to latency, such as is caused by
+link congestion, or the number of other agents delivering to the same
+sink, can cause the same settings to now impede utilization rather than
+improve it.
+
+Since many of these factors that affect delivery rate are not fixed
+values but instead will vary considerably throughout the life of a
+service, it is impossible to choose the "best" parameters that will fit
+all these conditions. Instead, Vector should adopt an approach that
+allows for dynamic adjustment of the underlying parameters based on the
+current conditions.
+
+When service limits are reached, Vector will experience a number of
+undesirable phenomenon, notably request latency, timeouts, and
+deferrals. These all decrease the overall flow rate while actually
+increasing actual bandwidth usage.
+
+## Guide-level Proposal
+
+There are two levels of controls we have in play—static and dynamic. The
+existing controls are static, and best describe service limits, such as
+maximum allowed request rates. What is needed is a set of dynamic
+controls that adapt to the underlying conditions and scale the service
+utilization on the fly.
+
+Since the controls under consideration are dependant on some form of
+queueing, the controls will be inserted at the same level as
+`TowerRequestSettings`. The existing rate limit controls will remain in
+order to provide a hard upper bound on service utilization (for example,
+to prevent over-use violations), but will be dynamically bounded by
+adjusting the concurrency.
+
+The underlying control will replace the `tower::limit::ConcurrencyLimit`
+layer with a new custom layer that dynamically adjusts the concurrency
+limit based on current conditions. It will track each request's result
+status (ie success, or deferral) and the round trip time (RTT). This
+will require a modified `tower::limit::Limit` structure that will add
+and remove permits as needed. A new `ResponseFuture` will forward the
+result of the request back to the invoking `ConcurrencyLimit` after a
+completed `poll` (in addition to the usual action of releasing the
+permit on `drop`).
+
+The algorithm used to control the limit will follow the AIMD framework:
+
+* The controller will maintain an moving average RTT of past requests
+  using an exponentially weighted moving average (EWMA). The weighting
+  (α) is to be experimentally determined.
+
+* The current response's RTT is compared to this moving average:
+
+  * If less than or equal to the average, the concurrency will be
+    incremented by one (additive increase) once per RTT, up to a maximum
+    of the in flight limit.
+
+  * If greater than the average, or the result was a failure of any
```

Comment:
I think it's worth clarifying that not all failures should result in a multiplicative decrease. `3XX`, `4XX`, and especially [`418`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/418) are not server-side errors.

Reply:
So noted